### DISCUSSION TITLE
Adding a node behind a proxy

### DISCUSSION THREAD

**Discussion by HighPriest** [gravitational/teleport#7943] [Q&A]
My architecture looks like this:
 - Public IP auth server
 - Tunneled proxy server in private network
 - Node inside this private network on a deeper, hidden network [connected to the proxy server on another NIC]

The Auth server and Proxy server show up correctly in the Node list.
But trying to add the Node on the hidden network keeps on erroring out on connection with `teleport.cluster.local`
Here are the details:
<details>
<summary>On the hidden NODE</summary>
x86_64 machine - 64bit OS - Ubuntu 20 LTS
<details>
<summary>Config</summary>
<pre>
<code>
teleport:
  nodename: NODE_NAME
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  auth_token: "TOKEN"
  ca_pin: "sha256"
  auth_servers:
    - 10.0.0.245:3080
    - teleport.my.domain:3025
auth_service:
  enabled: "no"
  listen_addr: 0.0.0.0:3025
  cluster_name: teleport.my.domain
  public_addr: teleport.my.domain:443
ssh_service:
  enabled: "yes"
  labels:
    env: example
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "no"
  listen_addr: 0.0.0.0:3023
  web_listen_addr: :443
  public_addr: teleport.my.domain:443
  https_keypairs: []
  acme:
    enabled: "yes"
    email: admin@my.domain
</code>
</pre>
</details>
Error:
<code>
<pre>
Node failed to establish connection to cluster: Post "https://10.0.0.245:3080/v1/webapi/host/credentials": remote error: tls: internal error, Get "https://teleport.cluster.local/v2/cacert": remote error: tls: internal error. service/connect.go:70
</code>
</pre>
</details>
<details>
<summary>On PROXY server</summary>
RaspberryPi 4 - 64bit OS - ARMv8
<details>
<summary>Config</summary>
<pre>
<code>
teleport:
  nodename: raspberrypi
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  auth_token: "TOKEN"
  ca_pin: "sha256"
  diag_addr: ""
  auth_servers:
    - teleport.my.domain:3025
auth_service:
  enabled: "no"
  listen_addr: 0.0.0.0:3025
  cluster_name: teleport.my.domain
ssh_service:
  enabled: "yes"
  labels:
    env: example
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:3080
  public_addr: teleport.my.domain:443
  tunnel_listen_addr: 0.0.0.0:3080
  tunnel_public_addr: teleport.my.domain:443
  https_keypairs: []
  acme:
    enabled: "yes"
    email: admin@my.domain
</code>
</pre>
</details>
Error:
<pre>
<code>
acme can&#39;t get a cert for domain teleport.cluster.local, add it to the proxy_service.public_addr, or use one of the domains: MY PUBLIC DOMAIN] multiplexer/web.go:137
</code>
</pre>
</details>

**Comment by webvictim** [comment]
What's the command you're using to start your node? I'm not 100% clear on your architecture - the usual model is that the proxy server is public and the auth server hidden rather than the other way round.

**Reply by HighPriest** [reply]
For both node and the proxy I launch and manage them with systemd `systemctl start teleport.service`, because they both have their config files present.
I've also tried `teleport start --roles=node,proxy` on the proxy, but it doesn't change a thing.

About the architecture I understand that a proxy, can relay information to an auth node.
Now, what I want to do is 'stack' proxies, so how I understand it, this is the logic I'm looking for
[Public IP] **Auth + Web client** -->
[Same Public IP] **Proxy**
--Firewall-->
[NATed network] **Proxy** -->
[Hidden network with no internet] **Node** I want to add to the Auth

**Reply by webvictim** [reply]
The proxy is actually the part that serves the web client rather than the auth server.

I think maybe you’re misunderstanding the purpose of proxy servers here - you don’t need to connect a proxy to a proxy with the intention of tunnelling (unless you’re trying to use Teleport’s trusted cluster architecture, in which case you’d need a full auth+proxy setup on your private network)

Or are you saying that the second proxy in your example is on the same machine as the first mention of a proxy but it’s multihomed i.e. it has two network interfaces with different IPs on different networks?

**Reply by HighPriest** [reply]
Ok, then can I create a cluster without a public interface/no public domains? 
And then connect it somehow to the cluster on the public node?

**Reply by webvictim** [reply]
There’s basically 3 options:

1) You could add a Teleport node which you use as an internal jump host server and connect over SSH to the machines behind the NAT (using `ssh -J` or `tsh ssh -J`). Teleport itself doesn’t directly support a concept of a proxy connected to a proxy - it only goes one layer deep.

2) Make outbound internet access available for each of your remote nodes back to your root Teleport proxy. This could be as simple as giving them outbound access to your Teleport proxy server’s IP/host and on port 443 only. You could also potentially use third party software to configure some kind of persistent inbound tunnel/VPN from the node with internet access that the remote nodes could then use to connect out.

3) If you want to keep track of a full network of machines which are all located behind a remote firewall (and manage access to them with RBAC etc) you’d probably be better off setting up a Teleport leaf cluster (auth and proxy server) on the remote network and joining it back to your root auth/proxy.

See our [trusted cluster documentation](https://goteleport.com/docs/trustedclusters/) for more details.
